diff --git a/iseq.c b/iseq.c
index a102941..6220aa8 100644
--- a/iseq.c
+++ b/iseq.c
@@ -1285,9 +1285,9 @@ rb_iseq_clone(VALUE iseqval, VALUE newcbase)
 	iseq1->orig = iseqval;
     }
     if (newcbase) {
-	iseq1->cref_stack = NEW_BLOCK(newcbase);
+	iseq1->cref_stack = NEW_NODE_LONGLIFE(NODE_BLOCK, newcbase, 0, 0);
 	if (iseq0->cref_stack->nd_next) {
-	    iseq1->cref_stack->nd_next = iseq0->cref_stack->nd_next;
+	    iseq1->cref_stack->nd_next = (NODE *)rb_gc_write_barrier((VALUE)iseq0->cref_stack->nd_next);
 	}
     }
 
diff --git a/vm_insnhelper.c b/vm_insnhelper.c
index 13bbcf7..0889e8a 100644
--- a/vm_insnhelper.c
+++ b/vm_insnhelper.c
@@ -1032,12 +1032,12 @@ static NODE *
 vm_cref_push(rb_thread_t *th, VALUE klass, int noex)
 {
     rb_control_frame_t *cfp = vm_get_ruby_level_caller_cfp(th, th->cfp);
-    NODE *cref = NEW_BLOCK(klass);
+    NODE *cref = NEW_NODE_LONGLIFE(NODE_BLOCK, klass, 0, 0);
     cref->nd_file = 0;
     cref->nd_visi = noex;
 
     if (cfp) {
-	cref->nd_next = vm_get_cref(cfp->iseq, cfp->lfp, cfp->dfp);
+	cref->nd_next = (NODE *)rb_gc_write_barrier((VALUE)vm_get_cref(cfp->iseq, cfp->lfp, cfp->dfp));
     }
 
     return cref;
diff --git a/vm_insnhelper.h b/vm_insnhelper.h
index 74ca845..838a652 100644
--- a/vm_insnhelper.h
+++ b/vm_insnhelper.h
@@ -154,9 +154,9 @@ extern VALUE ruby_vm_const_missing_count;
 
 #define COPY_CREF(c1, c2) do {  \
   NODE *__tmp_c2 = (c2); \
-  c1->nd_clss = __tmp_c2->nd_clss; \
-  c1->nd_visi = __tmp_c2->nd_visi; \
-  c1->nd_next = __tmp_c2->nd_next; \
+  c1->nd_clss = rb_gc_write_barrier((VALUE)__tmp_c2->nd_clss);\
+  c1->nd_visi = __tmp_c2->nd_visi;\
+  c1->nd_next = (NODE *)rb_gc_write_barrier((VALUE)__tmp_c2->nd_next);\
 } while (0)
 
 #define CALL_METHOD(num, blockptr, flag, id, mn, recv) do { \
